File: attribute-and-term-counts-drift/python/recount_terms.py
"""Recount WooCommerce attribute terms whose cached count drifted from the real
catalog. Cross-checks recent Stripe sales to flag drifted terms that are still
actively selling as higher priority. Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/woocommerce/attribute-and-term-counts-drift/
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("recount_terms")

stripe.api_key = os.environ.get("STRIPE_SECRET_KEY", "sk_test_dummy")
WOO_URL = os.environ.get("WOO_STORE_URL", "https://example.com").rstrip("/")
AUTH = HTTPBasicAuth(
    os.environ.get("WOO_CONSUMER_KEY", "ck_dummy"),
    os.environ.get("WOO_CONSUMER_SECRET", "cs_dummy"),
)
SALES_LOOKBACK_HOURS = int(os.environ.get("SALES_LOOKBACK_HOURS", "24"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def all_attributes():
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/products/attributes", auth=AUTH, timeout=30)
    r.raise_for_status()
    return r.json()


def attribute_terms(attribute_id):
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/products/attributes/{attribute_id}/terms",
            params={"per_page": 100, "page": page}, auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for term in batch:
            yield term
        page += 1


def real_count(attribute_slug, term_slug):
    r = requests.get(
        f"{WOO_URL}/wp-json/wc/v3/products",
        params={
            "attribute": attribute_slug,
            "attribute_term": term_slug,
            "status": "publish",
            "stock_status": "instock",
            "per_page": 1,
        },
        auth=AUTH, timeout=30,
    )
    r.raise_for_status()
    return int(r.headers.get("X-WP-Total", "0"))


def decide(term, real):
    """Pure decision: compare the term's stored count to the freshly computed
    real count and decide whether to repair it. No I/O, easy to unit test.
    """
    stored = term.get("count", 0)
    if stored == real:
        return ("skip", "count already correct")
    if real < 0:
        return ("skip", "real count invalid, will not write a negative number")
    return ("repair", f"stored {stored}, real {real}")


def write_count(attribute_id, term_id, real):
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/products/attributes/{attribute_id}/terms/{term_id}",
        json={"count": real}, auth=AUTH, timeout=30,
    ).raise_for_status()


def recently_sold_product_ids(lookback_hours):
    since = int(time.time()) - lookback_hours * 3600
    ids = set()
    for intent in stripe.PaymentIntent.list(limit=100, created={"gte": since}).auto_paging_iter():
        order_id = intent.metadata.get("order_id")
        if intent.status != "succeeded" or not order_id:
            continue
        r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
        if r.status_code != 200:
            continue
        for line in r.json().get("line_items", []):
            ids.add(line.get("product_id"))
    return ids


def run():
    repaired = 0
    sold_ids = recently_sold_product_ids(SALES_LOOKBACK_HOURS)
    for attribute in all_attributes():
        for term in attribute_terms(attribute["id"]):
            real = real_count(attribute["slug"], term["slug"])
            action, reason = decide(term, real)
            if action == "skip":
                continue
            urgent = bool(sold_ids) and real > 0
            log.info(
                "Term %s (%s): %s. %s%s",
                term["name"], attribute["name"], reason,
                "would repair" if DRY_RUN else "repairing",
                " [urgent: recent sales use this term]" if urgent else "",
            )
            if not DRY_RUN:
                write_count(attribute["id"], term["id"], real)
            repaired += 1
    log.info("Done. %d term(s) %s.", repaired, "to repair" if DRY_RUN else "repaired")


if __name__ == "__main__":
    run()

File: attribute-and-term-counts-drift/node/recount-terms.js
/**
 * Recount WooCommerce attribute terms whose cached count drifted from the real
 * catalog. Cross-checks recent Stripe sales to flag drifted terms that are still
 * actively selling as higher priority. Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/woocommerce/attribute-and-term-counts-drift/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const SALES_LOOKBACK_HOURS = Number(process.env.SALES_LOOKBACK_HOURS || 24);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function allAttributes() {
  return woo("/products/attributes");
}

async function* attributeTerms(attributeId) {
  let page = 1;
  while (true) {
    const batch = await woo(`/products/attributes/${attributeId}/terms?per_page=100&page=${page}`);
    if (!batch.length) return;
    for (const term of batch) yield term;
    page++;
  }
}

async function realCount(attributeSlug, termSlug) {
  const url = `${WOO_URL}/wp-json/wc/v3/products?attribute=${attributeSlug}` +
    `&attribute_term=${termSlug}&status=publish&stock_status=instock&per_page=1`;
  const res = await fetch(url, { headers: { Authorization: AUTH } });
  if (!res.ok) throw new Error(`Woo products lookup returned ${res.status}`);
  await res.json();
  return Number(res.headers.get("X-WP-Total") || 0);
}

/**
 * Pure decision: compare the term's stored count to the freshly computed real
 * count and decide whether to repair it. No I/O, easy to unit test.
 */
export function decide(term, real) {
  const stored = term.count || 0;
  if (stored === real) return ["skip", "count already correct"];
  if (real < 0) return ["skip", "real count invalid, will not write a negative number"];
  return ["repair", `stored ${stored}, real ${real}`];
}

async function writeCount(attributeId, termId, real) {
  await woo(`/products/attributes/${attributeId}/terms/${termId}`, {
    method: "PUT",
    body: JSON.stringify({ count: real }),
  });
}

async function recentlySoldProductIds(lookbackHours) {
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  const ids = new Set();
  for await (const intent of stripe.paymentIntents.list({ limit: 100, created: { gte: since } })) {
    const orderId = intent.metadata.order_id;
    if (intent.status !== "succeeded" || !orderId) continue;
    const order = await woo(`/orders/${orderId}`).catch(() => null);
    if (!order) continue;
    for (const line of order.line_items || []) ids.add(line.product_id);
  }
  return ids;
}

export async function run() {
  let repaired = 0;
  const soldIds = await recentlySoldProductIds(SALES_LOOKBACK_HOURS);
  for (const attribute of await allAttributes()) {
    for await (const term of attributeTerms(attribute.id)) {
      const real = await realCount(attribute.slug, term.slug);
      const [action, reason] = decide(term, real);
      if (action === "skip") continue;
      const urgent = soldIds.size > 0 && real > 0;
      console.log(
        `Term ${term.name} (${attribute.name}): ${reason}. ` +
        `${DRY_RUN ? "would repair" : "repairing"}${urgent ? " [urgent: recent sales use this term]" : ""}`
      );
      if (!DRY_RUN) await writeCount(attribute.id, term.id, real);
      repaired++;
    }
  }
  console.log(`Done. ${repaired} term(s) ${DRY_RUN ? "to repair" : "repaired"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
